### ISSUE TITLE
Adding domain support

### ISSUE THREAD

**Issue by TheCardiall** [netdata/netdata#4011]
Hi,
After several tests I realized that allow connections from don't support domain.
I think it could be a good feature !
What do you think ?

**Comment by paulfantom** [comment]
I don't understand. Could you elaborate?

**Comment by TheCardiall** [comment]
I think an example would be more telling
```
[web] 
    allow connections from = localhost srv1.example.com  srv2.example.com ....
```

**Comment by paulfantom** [comment]
But clients usually doesn't have a domain, so how should it work? I'm confused.

**Comment by TheCardiall** [comment]
I tested with this syntax and it does not work

**Comment by Ferroin** [comment]
@TheCardiall Your request is still a bit unclear.  Are you asking for DNS lookups to work for the connection whitelist?

@paulfantom That really depends on the network environment.  If you've got a AD setup or a wholly managed corporate network, it's actually pretty typical that client systems do have a domain, and even a reverse-DNS entry.

**Comment by paulfantom** [comment]
@Ferroin True, I forgot about rDNS. Nonetheless @TheCardiall example doesn't look like PTR records.

**Comment by Ferroin** [comment]
But it is consistent with how a lot of things allow hostnames, including FQDN's, for access control definitions and then do name resolution on those to determine what IP's to allow (or, alternatively, do reverse DNS on clients who connect and check the resulting PTR record against the list of allowed names).

**Comment by paulfantom** [comment]
Let's let @TheCardiall explain what is the issue :)

**Comment by TheCardiall** [comment]
My request is quite simple I just want a dns resolution because for configuration reasons it is easier to put a domain name than an IP. For example in case of migrating my Prometheus server.

**Comment by Ferroin** [comment]
OK, that makes sense (and actually would be rather nice to have).

**Comment by paulfantom** [comment]
I would agree only if there is/will be (I am still rediscovering netdata features) same protection based on IP since domain spoofing is pretty simple thing to do.

Just a note: this **shouldn't** be considered a security feature since it will be vulnerable to simple IP/domain spoofing. 

But indeed it could be useful as a filter.

**Comment by ktsaou** [comment]
Guys, any IP can be resolved to any domain (reverse DNS), without any cooperation from the domain owner. So, if I own IP 1.2.3.4, I can make it reverse-resolve as `www.google.com` without asking google to do it.

This means hostnames and domain names cannot be used access control - anyone will be able to bypass them easily.

The only solution of access control at this point, is IP patterns.

**Comment by paulfantom** [comment]
> The only solution of access control at this point, is IP patterns.

Which is also susceptible to IP spoofing. But this is a little bit harder to do than domain spoofing.

**Comment by Ferroin** [comment]
> This means hostnames and domain names cannot be used access control - anyone will be able to bypass them easily.

You don't have to do it as reverse-DNS though.  In fact, most applications that allow hostnames for access control _don't_ do reverse-DNS for the exact reason you outline (and also because that approach does not scale well).  They do a regular forward lookup of the name only once (when they parse the config), which is harder to spoof (you have to own the domain name, or have really good timing and intercept and modify traffic) and, if done using libc routines, also gives you the option of things like using /etc/hosts, or LDAP, or any of various secured and authenticated DNS lookup mechanisms, or any number of other things.

**Comment by ktsaou** [comment]
> Which is also susceptible to IP spoofing. But this is a little bit harder to do than domain spoofing.

Can you give an example? TCP cannot be IP spoofed. Only UDP is able to do this.

**Comment by ktsaou** [comment]
> They do a regular forward lookup of the name only once

hm... so this `*.domain.com` will never work. You will have to put there all the hosts...

**Comment by ktsaou** [comment]
One solution could be to:

1. reverse DNS the IP find the hostname
2. forward DNS the hostname to verify it is the same IP
3. use the hostname in the pattern match

I think this cannot be spoofed either.

**Comment by Ferroin** [comment]
TCP can be IP spoofed.  Captive portals and transparent proxies do it all the time.  The only requirement is that you have to control the only route between the victim host and the actual host owning the IP being spoofed, which is actually not really any different from UDP spoofing beyond single packets.

As far as securing it, keep in mind that:

1. Not everyone will inherently be using DNS for this kind of thing (they may be using NIS, or LDAP, or even just `/etc/hosts`), or they may be using DNS in such a way that it's functionally not possible to spoof  (for example, using DNSCrypt, or doing signed AD lookups).  Just adding a warning that it's not secure to use this with regular DNS unless you absolutely trust the network you're operating over should cover that case just fine while still allowing people who know what they're doing to do it securely.
2. Requiring a reverse lookup that matches the forward lookup is a reasonably reliable way of securing this further against spoofing (it can still be spoofed through traffic modification by an upstream router), but it also makes it very difficult for anybody outside of a localized network to use.  It's a serious pain in the arse to set up properly working reverse DNS and keep it in sync with forward lookups.  This will also auto-fail pretty much any connection from a person with a consumer internet plan in the US (being able to set up reverse DNS for your hosts is largely a corporate-only thing here), and will cause other issues for people dealing with ISP's who do NAT at their gateway (that is, large groups of customers share the same single external IP address, which is not unusual in developing countries because it's apparently still too difficult to properly deploy IPv6 20+ years after it was introduced).  Given this, most stuff checks that the forward and reverse lookups match, but has an option to disable this behavior.
3. Doing just one look up at startup gives far less opportunity for someone to spoof things than checking each client as it connects.  It also scales far better, and makes it far easier to audit that netdata is doing what it says.

As far as handling of domains instead of networks, there are two major options for this:

1. Use the libc name resolution routines.  If the call to `gethostbyname()` fails, try `getnetbyname()`, which will do a lookup in `/etc/networks` or various other (non-DNS) methods.  This is an easy option that will work trivially for the case of properly managed corporate networks, who are likely to be the primary consumers of things like this, and can be easily made to work for cases like small isolated networks (you just have to define the network in `/etc/networks`.
2. Do a reverse lookup for each client that connects, confirm that the forward lookup matches, and then check that the end of the FQDN for the connected client is under the domain.  This technically works, but it can lead to odd and confusing behavior if the reverse DNS name does not match what the user expects, and the forward name is still valid.  It also requires two DNS lookups for every client connection, which is horrible for performance.

**Comment by paulfantom** [comment]
Just my 3 cents, as a principle I don't agree with implementing features which later are described as a "security feature only when X", since this leads to a lot of confusion.

A lot of people are using external DNS server which isn't in their control. This leads to a scenario where DNS spoofing is possible. For all those people implementing domain-based ACL isn't a security feature. And if this isn't a security feature for everyone then I don't see a reason for implementing this and advertising it as one, since this is bad marketing.

DNS isn't a protocol which handles security well (it kind of does when some extensions are enabled), so why would you like to implement security features using insecure protocol? For me implementing domain-level ACL will end up like this:

![the_most_useless_things_you_will_ever_see_640_02](https://user-images.githubusercontent.com/3531758/43340334-4515532c-91dc-11e8-8f5a-48ef23e0c78d.jpg)

**Comment by Ferroin** [comment]
It's not a security feature at all, it's a convenience feature that happens to be less secure in some circumstances than others.  Put simply, best case scenario, it's no less secure than just using IP addresses directly, worst case it's marginally less secure (being able to spoof DNS properly generally also means you can spoof IP addresses just as easily).

Regardless of how secure it is though, it's a significant improvement in convenience.  Imagine having to change a client address for some reason and then having to update _all_ your netdata instances to reflect that change, versus just updating the DNS record and restarting all the netdata instances (and if you're on a properly set up AD managed network, you don't even need to update the DNS record, because the client system will do it automatically for you).

**Comment by Ferroin** [comment]
Updated previous comment on GitHub.

**Comment by ktsaou** [comment]
Closing due to inactivity.
It seems there is no demand for this request.
